rts/motoko-rts/src/idl.rs

```diff
@@ -1215,24 +1215,34 @@ unsafe extern "C" fn idl_sub_buf_init(rel_buf: *mut u32, typtbl_size1: u32, typt
 }
 
 #[ic_mem_fn]
-unsafe fn idl_sub<M: Memory>(
+unsafe fn idl_alloc_typtbl<M: Memory>(
     mem: &mut M,
+    candid_data: Value,
+    type_offsets: Value,
+    typtbl_out: *mut *mut *mut u8,
+    typtbl_end_out: *mut *mut u8,
+    typtbl_size_out: *mut u32,
+) {
+    let mut type_descriptor = TypeDescriptor::new(candid_data, type_offsets);
+    *typtbl_out = type_descriptor.build_type_table(mem);
+    *typtbl_end_out = type_descriptor.type_table_end();
+    *typtbl_size_out = type_descriptor.type_count() as u32;
+}
+
+#[no_mangle]
+unsafe extern "C" fn idl_sub(
     rel_buf: *mut u32, // a buffer with at least 2 * typtbl_size1 * typtbl_size2 bits
     typtbl1: *mut *mut u8,
+    typtbl2: *mut *mut u8,
     typtbl_end1: *mut u8,
+    typtbl_end2: *mut u8,
     typtbl_size1: u32,
-    candid_data2: Value,
-    type_offsets2: Value,
+    typtbl_size2: u32,
     t1: i32,
     t2: i32,
 ) -> bool {
     debug_assert!(rel_buf != (0 as *mut u32));
 
```

Comment:
TODO: assert typetbl_end1/_end2 not null

Reply:
Thanks for reminding. I added the corresponding safety check.